File: frontend/src/api/viajesLaborales.ts
import { fetchWithAuth } from "./http.ts";

export type ViajeLaboralEstado =
  | "borrador"
  | "pendiente"
  | "aprobado"
  | "rechazado"
  | "cancelado";

export type ViajeLaboralListItem = {
  id: number;
  empleado_id: number;
  empleado_nombre: string | null;
  numero_empleado: string | null;
  fecha_salida: string;
  fecha_regreso: string;
  lugar_origen: string;
  lugar_destino: string;
  motivo: string;
  descripcion: string | null;
  medio_transporte: string;
  hospedaje: string | null;
  viaticos_estimados: number | null;
  estado: ViajeLaboralEstado;
  registrado_por_id: number;
  registrado_por_nombre: string | null;
  aprobado_por_id: number | null;
  aprobado_por_nombre: string | null;
  motivo_rechazo: string | null;
  created_at: string;
  updated_at: string;
};

export type ViajesLaboralesPageResponse = {
  items: ViajeLaboralListItem[];
  total: number;
  page: number;
  page_size: number;
};

export type ViajeLaboralEstadoOption = {
  value: ViajeLaboralEstado;
  label: string;
};

export type ViajesLaboralesEstadisticasResponse = {
  total: number;
  pendientes: number;
  aprobados: number;
  cancelados: number;
};

export type ViajeLaboralPayload = {
  empleado_id: number;
  fecha_salida: string;
  fecha_regreso: string;
  lugar_origen: string;
  lugar_destino: string;
  motivo: string;
  descripcion?: string | null;
  medio_transporte: string;
  hospedaje?: string | null;
  viaticos_estimados?: number | null;
};

export type ViajesLaboralesListParams = {
  page: number;
  page_size: number;
  empleado_id?: number;
  fecha_inicio?: string;
  fecha_fin?: string;
  destino?: string;
  estado?: ViajeLaboralEstado | "";
  busqueda?: string;
};

export type ViajesLaboralesFetchError = { status: number; detail: string };

async function readErrorDetail(res: Response): Promise<string> {
  const raw = await res.text();
  try {
    const j = JSON.parse(raw) as { detail?: unknown };
    if (typeof j.detail === "string") return j.detail;
    if (Array.isArray(j.detail)) {
      return j.detail
        .map((item) => {
          if (typeof item === "string") return item;
          if (item && typeof item === "object" && "msg" in item) {
            return String((item as { msg: unknown }).msg);
          }
          return JSON.stringify(item);
        })
        .join("; ");
    }
  } catch {
    /* ignore */
  }
  return raw || res.statusText || "Error";
}

function buildListParams(params: ViajesLaboralesListParams): URLSearchParams {
  const sp = new URLSearchParams();
  sp.set("page", String(params.page));
  sp.set("page_size", String(params.page_size));
  if (params.empleado_id != null) sp.set("empleado_id", String(params.empleado_id));
  if (params.fecha_inicio?.trim()) sp.set("fecha_inicio", params.fecha_inicio.trim());
  if (params.fecha_fin?.trim()) sp.set("fecha_fin", params.fecha_fin.trim());
  if (params.destino?.trim()) sp.set("destino", params.destino.trim());
  if (params.estado) sp.set("estado", params.estado);
  if (params.busqueda?.trim()) sp.set("busqueda", params.busqueda.trim());
  return sp;
}

export async function getViajesLaboralesPage(
  params: ViajesLaboralesListParams,
): Promise<ViajesLaboralesPageResponse> {
  const res = await fetchWithAuth(
    `/api/v1/viajes-laborales?${buildListParams(params).toString()}`,
  );
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajesLaboralesPageResponse;
}

export async function getViajesLaboralesEstadisticas(
  params: Omit<ViajesLaboralesListParams, "page" | "page_size">,
): Promise<ViajesLaboralesEstadisticasResponse> {
  const sp = buildListParams({ ...params, page: 1, page_size: 1 });
  sp.delete("page");
  sp.delete("page_size");
  const res = await fetchWithAuth(`/api/v1/viajes-laborales/estadisticas?${sp.toString()}`);
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajesLaboralesEstadisticasResponse;
}

export async function getViajesLaboralesEstados(): Promise<ViajeLaboralEstadoOption[]> {
  const res = await fetchWithAuth("/api/v1/viajes-laborales/estados");
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  const data = (await res.json()) as { items: ViajeLaboralEstadoOption[] };
  return data.items;
}

export async function getViajeLaboral(id: number): Promise<ViajeLaboralListItem> {
  const res = await fetchWithAuth(`/api/v1/viajes-laborales/${id}`);
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajeLaboralListItem;
}

export async function createViajeLaboral(
  payload: ViajeLaboralPayload,
): Promise<ViajeLaboralListItem> {
  const res = await fetchWithAuth("/api/v1/viajes-laborales", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajeLaboralListItem;
}

export async function updateViajeLaboral(
  id: number,
  payload: Partial<ViajeLaboralPayload>,
): Promise<ViajeLaboralListItem> {
  const res = await fetchWithAuth(`/api/v1/viajes-laborales/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajeLaboralListItem;
}

export async function enviarViajeLaboral(id: number): Promise<ViajeLaboralListItem> {
  const res = await fetchWithAuth(`/api/v1/viajes-laborales/${id}/enviar`, { method: "PUT" });
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajeLaboralListItem;
}

export async function aprobarViajeLaboral(id: number): Promise<ViajeLaboralListItem> {
  const res = await fetchWithAuth(`/api/v1/viajes-laborales/${id}/aprobar`, { method: "PUT" });
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajeLaboralListItem;
}

export async function rechazarViajeLaboral(
  id: number,
  motivo_rechazo: string,
): Promise<ViajeLaboralListItem> {
  const res = await fetchWithAuth(`/api/v1/viajes-laborales/${id}/rechazar`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ motivo_rechazo }),
  });
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajeLaboralListItem;
}

export async function cancelarViajeLaboral(id: number): Promise<ViajeLaboralListItem> {
  const res = await fetchWithAuth(`/api/v1/viajes-laborales/${id}/cancelar`, { method: "PUT" });
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
  return (await res.json()) as ViajeLaboralListItem;
}

export async function deleteViajeLaboral(id: number): Promise<void> {
  const res = await fetchWithAuth(`/api/v1/viajes-laborales/${id}`, { method: "DELETE" });
  if (!res.ok) {
    throw { status: res.status, detail: await readErrorDetail(res) } as ViajesLaboralesFetchError;
  }
}

File: app/schemas/viajes_laborales.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, Field, model_validator

from app.models.viajes_laborales import VIAJE_LABORAL_ESTADOS

ViajeLaboralEstado = Literal[
    "borrador",
    "pendiente",
    "aprobado",
    "rechazado",
    "cancelado",
]

VIAJE_LABORAL_ESTADO_LABELS: dict[str, str] = {
    "borrador": "Borrador",
    "pendiente": "Pendiente",
    "aprobado": "Aprobado",
    "rechazado": "Rechazado",
    "cancelado": "Cancelado",
}


class ViajeLaboralCreate(BaseModel):
    model_config = {"str_strip_whitespace": True}

    empleado_id: int = Field(..., gt=0)
    fecha_salida: date
    fecha_regreso: date
    lugar_origen: str = Field(..., min_length=1, max_length=255)
    lugar_destino: str = Field(..., min_length=1, max_length=255)
    motivo: str = Field(..., min_length=1, max_length=4000)
    descripcion: Optional[str] = Field(None, max_length=4000)
    medio_transporte: str = Field(..., min_length=1, max_length=120)
    hospedaje: Optional[str] = Field(None, max_length=255)
    viaticos_estimados: Optional[Decimal] = Field(None, ge=0)

    @model_validator(mode="after")
    def validate_fechas(self) -> "ViajeLaboralCreate":
        if self.fecha_regreso < self.fecha_salida:
            raise ValueError("fecha_regreso no puede ser anterior a fecha_salida")
        return self


class ViajeLaboralUpdate(BaseModel):
    model_config = {"str_strip_whitespace": True}

    empleado_id: Optional[int] = Field(None, gt=0)
    fecha_salida: Optional[date] = None
    fecha_regreso: Optional[date] = None
    lugar_origen: Optional[str] = Field(None, min_length=1, max_length=255)
    lugar_destino: Optional[str] = Field(None, min_length=1, max_length=255)
    motivo: Optional[str] = Field(None, min_length=1, max_length=4000)
    descripcion: Optional[str] = Field(None, max_length=4000)
    medio_transporte: Optional[str] = Field(None, min_length=1, max_length=120)
    hospedaje: Optional[str] = Field(None, max_length=255)
    viaticos_estimados: Optional[Decimal] = Field(None, ge=0)

    @model_validator(mode="after")
    def validate_fechas(self) -> "ViajeLaboralUpdate":
        if (
            self.fecha_salida is not None
            and self.fecha_regreso is not None
            and self.fecha_regreso < self.fecha_salida
        ):
            raise ValueError("fecha_regreso no puede ser anterior a fecha_salida")
        return self


class ViajeLaboralRechazarRequest(BaseModel):
    model_config = {"str_strip_whitespace": True}

    motivo_rechazo: str = Field(..., min_length=1, max_length=4000)


class ViajeLaboralResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    empleado_id: int
    empleado_nombre: Optional[str] = None
    numero_empleado: Optional[str] = None
    fecha_salida: date
    fecha_regreso: date
    lugar_origen: str
    lugar_destino: str
    motivo: str
    descripcion: Optional[str] = None
    medio_transporte: str
    hospedaje: Optional[str] = None
    viaticos_estimados: Optional[Decimal] = None
    estado: ViajeLaboralEstado
    registrado_por_id: int
    registrado_por_nombre: Optional[str] = None
    aprobado_por_id: Optional[int] = None
    aprobado_por_nombre: Optional[str] = None
    motivo_rechazo: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class ViajesLaboralesPageResponse(BaseModel):
    items: list[ViajeLaboralResponse]
    total: int
    page: int
    page_size: int


class ViajeLaboralEstadoItem(BaseModel):
    value: ViajeLaboralEstado
    label: str


class ViajesLaboralesEstadosResponse(BaseModel):
    items: list[ViajeLaboralEstadoItem]


class ViajesLaboralesEstadisticasResponse(BaseModel):
    total: int
    pendientes: int
    aprobados: int
    cancelados: int


def list_estado_items() -> list[ViajeLaboralEstadoItem]:
    return [
        ViajeLaboralEstadoItem(value=estado, label=VIAJE_LABORAL_ESTADO_LABELS[estado])  # type: ignore[arg-type]
        for estado in VIAJE_LABORAL_ESTADOS
    ]
